File: day22/day22.cpp
#include <fstream>
#include <iostream>
#include <unordered_map>

#include <fmt/format.h>

using namespace std;

struct Vec
{
	int x;
	int y;

	bool operator==(const Vec& o) const noexcept
	{
		return x == o.x && y == o.y;
	}

	Vec& operator+=(const Vec& o) noexcept
	{
		x += o.x;
		y += o.y;
		return *this;
	}
};

static Vec left(Vec d)
{
	return Vec{d.y, -d.x};
}

static Vec right(Vec d)
{
	return Vec{-d.y, d.x};
}

static Vec reverse(Vec d)
{
	return Vec{-d.x, -d.y};
}

struct HashVec
{
	size_t operator()(const Vec& v) const noexcept
	{
		return (hash<int>{}(v.x) << 16)
			^ (hash<int>{}(v.y));
	}
};

using Map = unordered_map<Vec, int, HashVec>;

static Map make_map(const vector<string>& rows)
{
	Map m;
	for (int y = 0; (size_t)y < rows.size(); y++)
	{
		for (int x = 0; (size_t)x < rows[y].size(); x++)
		{
			if (rows[y][x] == '#')
			{
				m[Vec{x, y}] = '#';
			}
		}
	}
	return m;
}

static int part1(const vector<string>& initial, size_t rounds)
{
	Map m = make_map(initial);
	Vec p{(int)initial[0].size()/2, (int)initial.size()/2};
	Vec d{0, -1};
	size_t infections = 0;
	while (rounds-->0)
	{
		auto it = m.find(p);
		if (it != m.end())
		{
			m.erase(it);
			d = right(d);
		}
		else
		{
			m[p] = '#';
			d = left(d);
			infections++;
		}
		p += d;
	}
	return infections;
}

static int part2(const vector<string>& initial, size_t rounds)
{
	Map m = make_map(initial);
	Vec p{(int)initial[0].size()/2, (int)initial.size()/2};
	Vec d{0, -1};
	size_t infections = 0;
	while (rounds-->0)
	{
		auto it = m.find(p);
		if (it == m.end())
		{
			m[p] = 'W';
			d = left(d);
		}
		else if (it->second == 'W')
		{
			it->second = '#';
			infections++;
		}
		else if (it->second == '#')
		{
			it->second = 'F';
			d = right(d);
		}
		else if (it->second == 'F')
		{
			m.erase(it);
			d = reverse(d);
		}
		p += d;
	}
	return infections;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	vector<string> rows;
	string row;
	while (getline(input, row))
	{
		rows.emplace_back(move(row));
	}
	input.close();

	cout << fmt::format("Part1: {}", part1(rows, 10000ULL)) << endl
	     << fmt::format("Part2: {}", part2(rows, 10000000ULL)) << endl;
	return 0;
}

File: day22/day22.py
#!/usr/bin/env python3

import sys

def load_map(txt):
    m = {}
    rows = txt.splitlines()
    pos = (len(rows[0])//2,len(rows)//2)
    for y, row in enumerate(rows):
        for x, val in enumerate(row):
            if val == "#":
                m[x, y] = "#"

    return m, pos

def left(d):
    return (d[1], -d[0])

def right(d):
    return (-d[1], d[0])

def reverse(d):
    return (-d[0], -d[1])

def add(a, b):
    return (a[0]+b[0], a[1]+b[1])

def move(txt, rounds):
    m, p = load_map(txt)
    d = (0, -1)
    infections = 0
    for i in range(rounds):
        if p in m:
            d = right(d)
            del m[p]
        else:
            d = left(d)
            m[p] = "#"
            infections += 1

        p = add(p, d)
    return infections

def move2(txt, rounds):
    infections = 0
    m, p = load_map(txt)
    d = (0, -1)
    for i in range(rounds):
        value = m.get(p, ".")
        if value == ".":
            m[p] = "W"
            d = left(d)
        elif value == "W":
            m[p] = "#"
            infections += 1
        elif value == "#":
            m[p] = "F"
            d = right(d)
        elif value == "F":
            del m[p]
            d = reverse(d)

        p = add(p, d)
    return infections

def print_map(m, p):
    minx = miny = 1e12
    maxx = maxy = -1e12
    for x, y in m.keys():
        if minx > x: minx = x
        if maxx < x: maxx = x
        if miny > y: miny = y
        if maxy < y: maxy = y
    for y in range(miny, maxy+1):
        for x in range(minx, maxx+1):
            if (x, y) == p:
                print("@", end="")
            elif (x, y) in m:
                print("#", end="")
            else:
                print(".", end="")
        print()

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", move(txt, 10000))
    print("Part2:", move2(txt, 10000000))
